torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,89 +65,155 @@ def reset_table(self) -> None:
-    def visit_Call(self, node):
-        # TODO(oulgen): Here we assume that there exists a line called
-        # from triton import language as tl. This needs to be checked
-        # as if someones imports xyz as tl then we will incorrectly
-        # assume a mutation but this would be ok as it is only unsafe to
-        # miss a mutation.
-        if (
-            isinstance(node.func, ast.Attribute)
-            and isinstance(node.func.value, ast.Name)
-            and node.func.value.id == "tl"
-        ):
-            if node.func.attr == "store":
-                # Do not allow for store to appear inside a read
-                # tl.load(a if tl.store(b) else z) is not useful
-                # and allowing this would complicate the analysis
-                assert self.read_depth == 0
-                assert self.in_store is False
-                self.in_store = True
-                self.generic_visit(node)
-                self.in_store = False
-                return
-            if node.func.attr in self.ALLOWED_READ_FNS:
-                self.read_depth += 1
-                self.generic_visit(node)
-                self.read_depth -= 1
-                return
-        self.generic_visit(node)
-
-
-def filter_non_mutated(kernel, tensors):
+    def __add__(self, o):
+        if isinstance(o, Proxy):
+            raise IdentifyMutationException("Proxy operation with Proxy")
+        return self
+
+
+class Scalar:
+    @staticmethod
+    def check(o):
+        if isinstance(o, Proxy):
+            raise IdentifyMutationException("Proxy operation with Scalar")
+
+    def __radd__(self, o):
+        if isinstance(o, Proxy):
+            return o
+        return self
+
+    def __bool__(self):
+        raise IdentifyMutationException("Casting to bool is not allowed")
+
+
+# This operation is only needed when triton is available
+if has_triton():
+    import triton
+
+    def replacement_fn(self, *args, **kwargs):
+        if len(args) > 0:
+            Scalar.check(args[0])
+        return self
+
+    for name, _ in inspect.getmembers(triton.language.core.tensor, inspect.isfunction):
+        if name in ["__init__", "__radd__", "__bool__"]:
+            continue
+
+        setattr(Scalar, name, replacement_fn)
+
+
+# Given a triton kernel and the arguments for this kernel, this function traces
+# through the triton kernel and identifies which input pointers are mutated.
+# Tracing is done by replacing the input pointers with Proxy objects that
+# track mutation. Each triton language function is monkey patched to
+# either detect the mutation or return a fresh scalar object.
+def identify_mutated_tensors(kernel, kwargs):
+    from triton import language as tl
     from triton.runtime.autotuner import Autotuner
 
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
```

Comment:
Here you assume the argument is either a `Tensor` or a constant. Could we get a dynamic value such as a `SymInt` here?